### ISSUE TITLE
False-negative when instantiating a class before its constructor is defined

### ISSUE THREAD

**Issue by megapctr** [python/mypy#2324]
I've found a small bug:

```
class Operation:

    @classmethod
    def randomize(cls, config: Config) -> 'Operation':
        return cls(config)

    def __init__(self, config: Config) -> None:
        self.config = config
```

This complains at line 5, `Too many arugments for "Operation"`. Swapping these two methods around (init before randomize) fixed the issue.

Looks like mypy tried to type-check `cls(config)` before learning what `Operation`'s constructor looks like.

**Comment by gvanrossum** [comment]
Oh, interesting. This could be dealt with by deferring the first method when the class signature isn't known yet. But we'd need some kind of flag to tell us whether we've processed `__init__` yet (and maybe if there isn't one it should be set to True at the start, by the semantic analyzer).

Do you feel up to trying to submit a PR for this? No pressure!

**Comment by megapctr** [comment]
I could work on that, any tips what module/function to begin from?

**Comment by gvanrossum** [comment]
Best place to get help would be https://gitter.im/python/mypy.

**Comment by JukkaL** [comment]
Bumping priority because this seem to happen with some frequency and the workaround is not easy to figure out.

**Comment by JelleZijlstra** [comment]
This is a dupe of #1727, which is older and has more discussion; going to close this one and make the other one high-pri.

**Comment by mckaysalisbury** [comment]
It seems as if any method can call any method, so a two-pass system might work best. Once to process all of the function signatures, then another to process the code.

**Comment by gvanrossum** [comment]
Thanks for the advice.
